crates/ruff/src/rules/ruff/rules/implicit_optional.rs

```diff
@@ -0,0 +1,331 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::BinOpOr => f.write_str("T | None"),
+            Self::Optional => f.write_str("Optional[T]"),
+        }
+    }
+}
+
+impl From<PythonVersion> for ConversionType {
+    fn from(target_version: PythonVersion) -> Self {
+        if target_version >= PythonVersion::Py310 {
+            Self::BinOpOr
+        } else {
+            Self::Optional
+        }
+    }
+}
+
+/// Custom iterator to collect all the `|` separated expressions in a PEP 604
+/// union type.
+struct PEP604UnionIterator<'a> {
+    stack: Vec<&'a Expr>,
+}
+
+impl<'a> PEP604UnionIterator<'a> {
+    fn new(expr: &'a Expr) -> Self {
+        Self { stack: vec![expr] }
+    }
+}
+
+impl<'a> Iterator for PEP604UnionIterator<'a> {
+    type Item = &'a Expr;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(expr) = self.stack.pop() {
+            match expr {
+                Expr::BinOp(ast::ExprBinOp {
+                    left,
+                    op: Operator::BitOr,
+                    right,
+                    ..
+                }) => {
+                    self.stack.push(left);
+                    self.stack.push(right);
+                }
+                _ => return Some(expr),
+            }
+        }
+        None
+    }
+}
+
+#[derive(Debug)]
+enum TypingTarget<'a> {
+    None,
+    Optional,
+    Union(Vec<&'a Expr>),
+    Literal(Vec<&'a Expr>),
+    Annotated(&'a Expr),
+}
+
+impl<'a> TypingTarget<'a> {
+    fn try_from_expr(model: &SemanticModel, expr: &'a Expr) -> Option<Self> {
+        match expr {
+            Expr::Subscript(ast::ExprSubscript { value, slice, .. }) => {
+                if model.match_typing_expr(value, "Optional") {
+                    return Some(TypingTarget::Optional);
+                }
+                let Expr::Tuple(ast::ExprTuple { elts: elements, .. }) = slice.as_ref() else{
+                    return None;
+                };
+                if model.match_typing_expr(value, "Literal") {
+                    Some(TypingTarget::Literal(elements.iter().collect()))
+                } else if model.match_typing_expr(value, "Union") {
+                    Some(TypingTarget::Union(elements.iter().collect()))
+                } else if model.match_typing_expr(value, "Annotated") {
+                    elements.first().map(TypingTarget::Annotated)
+                } else {
+                    None
+                }
+            }
+            Expr::BinOp(..) => Some(TypingTarget::Union(
+                PEP604UnionIterator::new(expr).collect(),
+            )),
+            Expr::Constant(ast::ExprConstant {
+                value: Constant::None,
+                ..
+            }) => Some(TypingTarget::None),
+            _ => None,
+        }
+    }
+
+    /// Check if the [`TypingTarget`] explicitly allows `None`.
+    fn is_implicit_optional(&self, model: &SemanticModel) -> bool {
```

Comment:
Do we prefer to use `Self` or explicit type name `TypingTarget`?

Reply:
That makes sense. Without editor support, it might be hard to figure out what `Self` is.